### ISSUE TITLE
Miden VM executable and crate organization

### ISSUE THREAD

**Issue by bobbinth** [maticnetwork/miden#262]
Currently, two crates generate executable files: `examples` and `tools`. Both use functionality exposed by the `miden` crate, which in turn uses `processor` crate and `verifier` crate.

This seems like a fairly confusing arrangement, and it will probably get even more confusing as we add more things (e.g., repl tool). So, I'd like to see if we can re-shuffle things a bit to make the structure more intuitive.

One option is to do something like this:
1. Create a `prover` crate which will expose a `prove()` function. This function will execute programs and generate proofs of their execution (basically the same as the current `execute()` function in the `miden` crate). This will be consistent with having the `verifier` crate.
2. Change `miden` crate to output an executable. This executable would be used to run Miden VM from the command line in various modes. This will also allow us to get rid of `tools` create and probably `examples` crate as well.

The potential modes of `miden` executable could be as follows:
- `miden run` - execute a program and generate an output
- `miden prove` - execute a program and generate a proof
- `miden verify` - verify a previously generated proof
- `miden analyze` - run a script analysis tool
- `miden repl` - start Miden REPL

We still need to figure out how to provide inputs and read outputs for each mode. For example, for `miden prove` we could do something like this:
```
miden prove [path to .masm script file] -i [path to an input file] -n [number of outputs] -p [proof options] -o [output file name]
```
But much more thinking is needed about how to structure these.

**Comment by frisitano** [comment]
I can pick this up @bobbinth

**Comment by bobbinth** [comment]
@frisitano - that would be awesome! but i think as the first step, we should probably flash out what the parameters for each would look like. let me know if you'd like to take a stab at this (we can do it in this thread).

**Comment by frisitano** [comment]
Hey @bobbinth , yeah I'm happy to make an initial proposal for cli interface and then we can review?

**Comment by bobbinth** [comment]
@frisitano - sounds good! Thank you!

**Comment by frisitano** [comment]
I was thinking that something like this could work:
```
```

**Comment by frisitano** [comment]
I was thinking that something like this could work:
```
Miden.

Usage:
  miden run <script_file> <input_file> <num_outputs> <output_file>
  miden prove <script_file> <input_file> <num_outputs> <proof_file>  [--options]
  miden verify <script_file> <input_file> <proof_file>
  miden analyze <script_file> <input_file>
  miden repl TBD

Options:
  -h --help     show this screen
  --options     proof options [default: 96 bit security]
```

**Comment by bobbinth** [comment]
I like it! A few questions/comments:

- Should we make some of the parameters optional or default to something? For example:
  - For inputs, we could set the default to `script_file.inputs` (or something similar).
  - For outputs, the default could be just to print to stdout.
  - For `num_outputs` we could set the default to 16.
  - For proof file, we could set the default to `script_file.proof` (or something similar).
- Input files for prove and verify would have to be slightly different. For `verify` we need only the public inputs (i.e., initial values for the stack). For `prove` we also need values for populating the advice provider. Though, we probably don't need to tackle the advice provider immediately.
- Also, `verify`:
  - We don't need the full script file - providing just the hash should suffice.
  - We do need the output file as the proof does not contain output values.
- Any thoughts on the format for the input/output files (e.g. JSON/TOML/something else)?

**Comment by frisitano** [comment]
> I like it! A few questions/comments:
> 
> * Should we make some of the parameters optional or default to something? For example:
>   
>   * For inputs, we could set the default to `script_file.inputs` (or something similar).
>   * For outputs, the default could be just to print to stdout.
>   * For `num_outputs` we could set the default to 16.
>   * For proof file, we could set the default to `script_file.proof` (or something similar).

Yeah I think this makes sense.

> * Input files for prove and verify would have to be slightly different. For `verify` we need only the public inputs (i.e., initial values for the stack). For `prove` we also need values for populating the advice provider. Though, we probably don't need to tackle the advice provider immediately.

If we use JSON / TOML the input file can contain both public inputs and advice provider data - the respective command can extract the data required?

> * Also, `verify`:
>   
>   * We don't need the full script file - providing just the hash should suffice.

Maybe we provide the option of specifying the script file or alternatively a hash via the command line?  I'm just thinking of the UX where a user is forced to generate a hash to execute the verify command?

>   * We do need the output file as the proof does not contain output values.

My thinking was that the proof file could also contain the outputs as we need to generate the execution trace in any case.  However maybe it would be cleaner to have them separate?

> * Any thoughts on the format for the input/output files (e.g. JSON/TOML/something else)?

I'm not strongly opinionated - I think JSON, TOML, YAML are all fine.  Happy  to take your lead on this.

**Comment by bobbinth** [comment]
> If we use JSON / TOML the input file can contain both public inputs and advice provider data - the respective command can extract the data required?

Yes - agreed.

> Maybe we provide the option of specifying the script file or alternatively a hash via the command line?

I think both could be useful, but I think we should start with the second option. We could probably add another subcommand which takes a script and outputs its hash. Maybe `miden compile` or something like that?

> My thinking was that the proof file could also contain the outputs as we need to generate the execution trace in any case.

I have two reservations about combining them:
1. If we combine proof and outputs, why do we exclude the inputs? Maybe in this case we should put public inputs, outputs, and proof in the same file - but then it is no longer just a proof.
2. Proof is just a bunch of binary data, and we put it into a file like JSON, we'll need to do a base64 encoding, and that will increase the proof size by 33%.

> I think JSON, TOML, YAML are all fine.

Yeah - I don't have a strong preference either. For portability reasons, JSON is probably a little better.

**Comment by frisitano** [comment]
> I think both could be useful, but I think we should start with the second option. We could probably add another subcommand which takes a script and outputs its hash. Maybe `miden compile` or something like that?

This sounds reasonable.  What behaviour would we desire from the `miden compile` command.  Would we just want to print the hash stdout?

I'll start drafting some code.

**Comment by bobbinth** [comment]
> This sounds reasonable. What behaviour would we desire from the `miden compile` command. Would we just want to print the hash stdout?

Yes, for now it can probably just print out hash + time it took to compile. We can probably add more stats in the future.

**Comment by frisitano** [comment]
Tracking:

- [ ] add prover crate and move relevant functionality from miden crate there.
- [ ] merge examples into the miden crate and make miden crate output an executable (which would be the same one as currently produced by examples crate).
- [ ] refactor CLI interface of the miden crate and get rid of the tools crate (this can also be split up into a couple of PRs).

**Comment by bobbinth** [comment]
Closed by #285 and others.
